{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "018cf3d2",
   "metadata": {},
   "source": [
    "# BIOLOGICAL FEATURES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61c3e034",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ================================\n",
    "#           libraries\n",
    "# ================================\n",
    "\n",
    "# %pip install -q MDTraj    # leave this uncommented if running it on Google Colab\n",
    "import mdtraj as md\n",
    "from Bio import SeqIO\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.ndimage as nd\n",
    "import math\n",
    "from tqdm import tqdm\n",
    "from random import random\n",
    "import plotly.graph_objects as go\n",
    "import plotly.io as pio\n",
    "\n",
    "import urllib.request\n",
    "import ssl\n",
    "from copy import deepcopy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b79634b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://calla.rnet.missouri.edu/genome3d/GSDB/Database/AX9716PF/GSE105544_ENCFF010WBP/VC/LorDG/chr1.pdb'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "362ff408",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create unverified context for this specific download\n",
    "ssl_context = ssl._create_unverified_context()\n",
    "\n",
    "# Download the file\n",
    "with urllib.request.urlopen(url, context=ssl_context) as response:\n",
    "    with open('chr1.pdb', 'wb') as out_file:\n",
    "        out_file.write(response.read())\n",
    "\n",
    "# Load from local file\n",
    "chr_str = md.load_pdb('chr1.pdb')\n",
    "coords = chr_str.xyz.reshape(-1, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "610d6ab5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "____________________________________________________________\n",
      "ID: CM000663.2\n",
      "Description: CM000663.2 Homo sapiens chromosome 1, GRCh38 reference primary assembly\n",
      "____________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "fasta_file = '../data/raw/hg38_chr1_FASTA.fasta'\n",
    "\n",
    "sequences = []\n",
    "for record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "    sequences.append(record)\n",
    "    print('_' * 60)\n",
    "    print(f'ID: {record.id}')\n",
    "    # print(f'Sequence: {record.seq}')\n",
    "    print(f'Description: {record.description}')\n",
    "    print('_' * 60)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8a6a1832",
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_content = {'bin': [], 'gc_content': []}\n",
    "\n",
    "def gc_content_calculator(sequence, window):\n",
    "    for i in range(0, len(sequence) - window, window):\n",
    "        sequence_frag = sequence[i, i + window].upper()\n",
    "        g_content = sequence_frag.count('G')\n",
    "        c_content = sequence_frag.count('C')\n",
    "        gc_content['bin'].append(i)\n",
    "        gc_content['gc_content'].append((g_content + c_content)/ window)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "57609a8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGCG'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = 'ATGCGCCGATATAGCGGGGCGTGGAG'\n",
    "\n",
    "seq[0:5]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
