{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "018cf3d2",
   "metadata": {},
   "source": [
    "# BIOLOGICAL FEATURES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61c3e034",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ================================\n",
    "#           libraries\n",
    "# ================================\n",
    "\n",
    "# %pip install -q MDTraj    # leave this uncommented if running it on Google Colab\n",
    "# Standard library imports\n",
    "import math\n",
    "import ssl\n",
    "import urllib.request\n",
    "from copy import deepcopy\n",
    "from random import random\n",
    "\n",
    "# Third-party imports\n",
    "import mdtraj as md\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import plotly.graph_objects as go\n",
    "import plotly.io as pio\n",
    "import scipy.ndimage as nd\n",
    "from Bio import SeqIO\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b79634b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://calla.rnet.missouri.edu/genome3d/GSDB/Database/AX9716PF/GSE105544_ENCFF010WBP/VC/LorDG/chr1.pdb'\n",
    "# chromosome model used: https://gsdb.mu.hekademeia.org/details.php?id=GSE105544_ENCFF010WBP\n",
    "# resolution: 500Kb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d11e855c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading PDB file to check the raw file for some EDA\n",
    "\n",
    "with open ('chr1.pdb', 'r') as f:\n",
    "    file = f.read()\n",
    "\n",
    "lines = file.split('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "362ff408",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create unverified context for this specific download\n",
    "ssl_context = ssl._create_unverified_context()\n",
    "\n",
    "# Download the file\n",
    "with urllib.request.urlopen(url, context=ssl_context) as response:\n",
    "    with open('chr1.pdb', 'wb') as out_file:\n",
    "        out_file.write(response.read())\n",
    "\n",
    "# Load from local file\n",
    "chr_str = md.load_pdb('chr1.pdb')\n",
    "coords = chr_str.xyz.reshape(-1, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "610d6ab5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "____________________________________________________________\n",
      "ID: CM000663.2\n",
      "Description: CM000663.2 Homo sapiens chromosome 1, GRCh38 reference primary assembly\n",
      "____________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "fasta_file = '../data/raw/hg38_chr1_FASTA.fasta'\n",
    "\n",
    "sequences = []\n",
    "for record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "    sequences.append(record)\n",
    "    print('_' * 60)\n",
    "    print(f'ID: {record.id}')\n",
    "    # print(f'Sequence: {record.seq}')\n",
    "    print(f'Description: {record.description}')\n",
    "    print('_' * 60)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a6a1832",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gc_content_calculator(sequence, window) -> DataFrame:\n",
    "    \"\"\"Calculate GC content based on the window size\"\"\"\n",
    "\n",
    "    sequence_upper = str(sequence).upper()\n",
    "    seq_len = len(sequence_upper)\n",
    "    num_windows = (seq_len - window) // window + 1\n",
    "    bins = []\n",
    "    gc_percentages = []\n",
    "    bins.reserve = num_windows  # Hint for list allocation\n",
    "    gc_percentages.reserve = num_windows\n",
    "\n",
    "    gc_chars = {'G', 'C'}\n",
    "    for i in range(0, seq_len - window + 1, window):\n",
    "        sequence_frag = sequence_upper[i:i + window]\n",
    "        gc_count = sum(1 for char in sequence_frag if char in gc_chars)\n",
    "        if gc_count == 0 and 'N' in sequence_frag:\n",
    "            bins.append(i)\n",
    "            gc_percentages.append('telo-centro')\n",
    "        elif gc_count > 0:\n",
    "            gc_content_percentage = round((gc_count / window) * 100, 2)\n",
    "            bins.append(i)\n",
    "            gc_percentages.append(gc_content_percentage)\n",
    "\n",
    "    return pd.DataFrame({\n",
    "        'bin': bins,\n",
    "        'gc_content_percentage': gc_percentages\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "57609a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gc = np.array(gc_content_calculator(sequences[0].seq, 500000))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d7ae075",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ==========================================================\n",
    "#                       BINS STAT\n",
    "# ==========================================================\n",
    "# PDB ---> chr_str ---> 451 bins\n",
    "# FASTA -> total -----> 492 bins\n",
    "# FASTA -> w/o tel ---> 462 bins\n",
    "# =========================================================="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "4e95043b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(451, 5)"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_file = []\n",
    "for coord, g in zip(coords, gc):\n",
    "    merged_file.append([*coord, *g])\n",
    "\n",
    "merged_file = np.array(merged_file)\n",
    "merged_file.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "463c2293",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('x_y_z_bin_gc.npy', merged_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
